pkg/client/client.go

```diff
@@ -347,6 +348,21 @@ func (c *client) Patch(ctx context.Context, obj Object, patch Patch, opts ...Pat
 	}
 }
 
+func (c *client) Apply(ctx context.Context, obj Object, fieldOwner string) error {
```

Comment:
Pretty heavily opposed to do anything in here before https://github.com/kubernetes/kubernetes/issues/118138 is done because:
* The only type that can be used in here is `unstructured.Unstructured`
* The converting as you do it down below doesn't fix the issue that the normal types end up putting zero values in fields
* Which is the reason why applytypes exist, but you can not actually use applytypes here, because they do not fulfill `client.Object`
/hold

Reply:
Definitely hold, and no strong opinions on the feedback beyond understanding the gaps that exists. Waiting on this make sense and provided the support going forward, strategy around this seems crucial.  Here for that conversation.

Reply:
The whole topic boils down to "someone needs to do the work":
1. The upstream issue I mentioned needs fixing, upstream signaled that they are ok with the idea but no one has shown up to actually put the work in
2. This PR needs to be finished: https://github.com/kubernetes-sigs/controller-tools/pull/818
3. A change like this in controller-runtime that adds an `Apply` method that only accepts apply types